### PR TITLE
patch flang runtime library

### PR DESCRIPTION
**Pull request by xoviat**
patch the runtime library so that it works on windows.

### CODE REVIEW

File: runtime/flang/curdir.c
Comment (bryanpkc) [inline runtime/flang/curdir.c:108]: There should be an equivalent for `gethostname` in the Win32 or Winsock API, right?

Can you keep the support for `-curhost` on Windows?

Comment (xoviat) [reply runtime/flang/curdir.c:108]: There is, but I haven't put it in yet.

Comment (xoviat) [reply runtime/flang/curdir.c]: that should be good for now; I will add support for this later.

**Comment by xoviat** [comment]
@kaadam Are you going to look at this?

File: runtime/flang/hand.c
Comment (kaadam) [inline runtime/flang/hand.c:126]: maybe leave a 'fixme' or 'todo' comment?

File: runtime/flang/stat_linux.c
Comment (kaadam) [inline runtime/flang/stat_linux.c:12]: Since definition part of functions guarded maybe leave a fixme or todo comment here? Should we support it for Windows?

File: runtime/include/mthdecls.h
Comment (xoviat) [reply runtime/include/mthdecls.h:8]: Like this?

```suggestion
#if !defined(_WIN32) && !defined(HOST_MINGW)
```

File: runtime/flang/hand.c
Comment (xoviat) [reply runtime/flang/hand.c:129]:
```suggestion
#endif

```

File: runtime/flang/mmcmplx16.c
Comment (xoviat) [reply runtime/flang/mmcmplx16.c:550]:
```suggestion
}

```

File: runtime/flang/chdir3f.c
Comment (xoviat) [reply runtime/flang/chdir3f.c:14]: It compiles for me, but I guess I missed it.

File: runtime/flang/getpid3f.c
Comment (xoviat) [reply runtime/flang/getpid3f.c:13]: It compiles for me, but I guess I missed it.

File: runtime/flang/hand.c
Comment (xoviat) [reply runtime/flang/hand.c:126]: will do.

File: runtime/flang/stat_linux.c
Comment (xoviat) [reply runtime/flang/stat_linux.c:12]: will do. yes, we should support it, but mainly focused on getting to basic functionality now.

File: runtime/include/mthdecls.h
Comment (kaadam) [reply runtime/include/mthdecls.h:8]: Yes, that's perfect.  I think we should avoid superfluous macros.

File: runtime/flang/chdir3f.c
Comment (xoviat) [inline runtime/flang/chdir3f.c:15]:
```suggestion
#else
#include <direct.h>
#endif
```

File: runtime/flang/getpid3f.c
Comment (xoviat) [inline runtime/flang/getpid3f.c:15]:
```suggestion
#else
#include <process.h>
#endif
```

File: runtime/flang/hand.c
Comment (xoviat) [reply runtime/flang/hand.c:126]:
```suggestion
// TODO: implement __fort_sethand
__fort_sethand()
```

File: runtime/flang/stat_linux.c
Comment (xoviat) [inline runtime/flang/stat_linux.c:35]:
```suggestion
// TODO: Implement functions for WIN32
```

**Review by kaadam — APPROVED** [review]
LGTM, nit: Could you add a description about what you have done this patch? Thanks.

**Comment by pawosm-arm** [comment]
Could you squash those 18 commits into one?

**Comment by xoviat** [comment]
@pawosm-arm do I need feedback from shivaramaarao first? Or do you have write access?

**Comment by kiranchandramohan** [comment]
@shivaramaarao has to approve for the x86_64/amd64 platform. @pawosm-arm, @bryanpkc or myself can approve for aarch64.

**Comment by xoviat** [comment]
I'll wait until he approves, then squash them.

**Comment by kaadam** [comment]
@bryanpkc @shivaramaarao @kiranchandramohan Could you please take a look at this PR? Thanks in advance!

File: runtime/include/mthdecls.h
Comment (bryanpkc) [inline runtime/include/mthdecls.h:9]: This should be defined to `float complex`.

File: runtime/include/mthdecls.h
Comment (bryanpkc) [inline runtime/include/mthdecls.h:14]: This should be defined to `double complex`.

Comment (xoviat) [reply runtime/include/mthdecls.h:14]:
```suggestion
#define DOUBLE_COMPLEX_TYPE double complex
```

**Comment by xoviat** [comment]
@kiranchandramohan is  there any chance that this will be merged soon?

File: runtime/flang/async.c
Comment (kiranchandramohan) [inline runtime/flang/async.c:298]: Can your provide a pointer to the OVERLAPPED structure? Is this from a windows header?

File: runtime/flang/buffer.c
Comment (kiranchandramohan) [inline runtime/flang/buffer.c:12]: What is PARAMID here?

File: runtime/flang/heapinit.c
Comment (kiranchandramohan) [inline runtime/flang/heapinit.c:36]: What will happen in windows?

File: runtime/include/mthdecls.h
Comment (kiranchandramohan) [inline runtime/include/mthdecls.h:8]: What will happen in Windows?

File: runtime/flang/stat_linux.c
Comment (kiranchandramohan) [inline runtime/flang/stat_linux.c:36]: Can you file a github issue listing all the TODOs for windows?

**Review by kiranchandramohan — COMMENTED** [review]
A few quick comments.

File: runtime/flang/xfer_heap_dum.c
Comment (xoviat) [reply runtime/flang/xfer_heap_dum.c:19]: It's a compilation fix for now. These functions can be implemented later once initial windows functionality is working.

File: runtime/flang/alias.c
Comment (xoviat) [reply runtime/flang/alias.c:12]: I don't think the braces are consistent. Would you like me to fix that?

File: runtime/flang/async.c
Comment (xoviat) [reply runtime/flang/async.c:298]: `.u.s` doesn't work with windows. It throws a compilation error.

File: runtime/flang/buffer.c
Comment (xoviat) [reply runtime/flang/buffer.c:12]: https://www.pyramidanalytics.com/?

File: runtime/flang/heapinit.c
Comment (xoviat) [reply runtime/flang/heapinit.c:36]: We don't need to handle the signal on windows, since it doesn't exist.

File: runtime/include/mthdecls.h
Comment (xoviat) [reply runtime/include/mthdecls.h:8]: The macros are defined below.

File: runtime/flang/stat_linux.c
Comment (xoviat) [reply runtime/flang/stat_linux.c:36]: I don't know what all of the TODOs are for windows. I know that I can use the compiler without this functionality without any issues.

File: runtime/include/mthdecls.h
Comment (kiranchandramohan) [reply runtime/include/mthdecls.h:8]: Ahh sorry, I missed the else in between.

File: runtime/flang/xfer_heap_dum.c
Comment (kiranchandramohan) [reply runtime/flang/xfer_heap_dum.c:19]: Can you add a TODO in all these places? Otherwise how will you find these later?

File: runtime/flang/alias.c
Comment (kiranchandramohan) [reply runtime/flang/alias.c:12]: Keep the braces and spacing (2) consistent.

File: runtime/flang/buffer.c
Comment (kiranchandramohan) [reply runtime/flang/buffer.c:12]: Did you get the link wrong?

File: runtime/flang/async.c
Comment (kiranchandramohan) [reply runtime/flang/async.c:298]: OK. I didn't see a definition in flang. Assuming it is from a windows header.

**Comment by kiranchandramohan** [comment]
I will run a few tests tomorrow and if OK will accept. Does @bryanpkc have further comments?

File: runtime/flang/xfer_heap_dum.c
Comment (kiranchandramohan) [reply runtime/flang/xfer_heap_dum.c:19]: I am approving but I think this is important to keep track of where further implementation is required.

**Review by kiranchandramohan — APPROVED** [review]
I would like to merge this PR before this Wednesday's general Flang call if there are no further comments.

File: runtime/flang/xfer_heap_dum.c
Comment (kiranchandramohan) [reply runtime/flang/xfer_heap_dum.c:19]: Will add TODO in a separate patch.

File: runtime/flang/curdir.c
Comment (kiranchandramohan) [reply runtime/flang/curdir.c]: Will add TODO in a separate patch.

**Comment by bryanpkc** [comment]
Sorry I missed the messages on this PR. Thanks for merging @kiranchandramohan.

**Comment by bryanpkc** [comment]
@xoviat  For future reference please following the coding convention and use 2-space indents.

**Pull request merged**
